src/Datasource/Connection.php

```diff
@@ -88,69 +108,117 @@ public function __construct(array $config = [], $callback = null)
      */
     public function __call($name, $attributes)
     {
-        if (method_exists($this->_client, $name)) {
-            return call_user_func_array([$this->_client, $name], $attributes);
+        $client = $this->getDriver()->getConnection();
+        if (method_exists($client, $name)) {
+            return call_user_func_array([$client, $name], $attributes);
         }
     }
 
     /**
-     * Returns a SchemaCollection stub until we can add more
-     * abstract API's in Connection.
-     *
-     * @return \Cake\ElasticSearch\Datasource\SchemaCollection
+     * @inheritDoc
      */
-    public function getSchemaCollection()
+    public function config()
     {
-        return new SchemaCollection($this);
+        return $this->_config;
     }
 
     /**
-     * {@inheritDoc}
+     * @inheritDoc
      */
     public function configName()
     {
-        return $this->configName;
+        if (empty($this->_config['name'])) {
+            return '';
+        }
+
+        return $this->_config['name'];
     }
 
     /**
-     * {@inheritDoc}
+     * Sets the driver instance. If a string is passed it will be treated
+     * as a class name and will be instantiated.
+     *
+     * @param \Cake\ElasticSearch\Database\Driver|string $driver The driver instance to use.
+     * @param array $config Config for a new driver.
+     * @throws \Cake\Database\Exception\MissingDriverException When a driver class is missing.
+     * @return $this
      */
-    public function enabled()
+    public function setDriver($driver, $config = [])
     {
-        return true;
+        if (is_string($driver)) {
+            $className = App::className($driver, 'Database/Driver');
+            if (!$className || !class_exists($className)) {
+                throw new MissingDriverException(['driver' => $driver]);
+            }
+
+            $config['esLogger'] = $this->getEsLogger();
+            $callback = $config['callback'] ?? null;
+            $driver = new $className($config, $callback);
```

Comment:
What would happen if this classname wasn't elasticsearch based? Perhaps this should be constrained to subclasses of the elasticsearch driver class for now?

Reply:
This is fair, I'll put some validation there.